### PR TITLE
[BugFix] jitExpr must unregister jit_function only once

### PR DESCRIPTION
**Pull request by fzhedu**
Why I'm doing:

When pushing down runtime filter in Project, the expr should additionally call prepare-open-eval-close. as Prepared is called just once to register jit_function, so the jit_function must unregister once at last.

What I'm doing:

 jitExpr must unregister jit_function only once, so it cann't do in close(), instead in ~jitExpr().

Fixes https://github.com/StarRocks/StarRocksTest/issues/5597

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/jit/jit_expr.cpp [hunk 1]
```diff
@@ -121,16 +115,17 @@ StatusOr<ColumnPtr> JITExpr::evaluate_checked(starrocks::ExprContext* context, C
...
+// only unregister once
+JITExpr::~JITExpr() {
+    if (_is_prepared && _jit_function != nullptr) {
+        auto* jit_engine = JITEngine::get_instance();
+        if (jit_engine->initialized()) {
+            auto status = jit_engine->remove_function(_expr->debug_string());
+            if (!status.ok()) {
+                LOG(WARNING) << "JIT: remove function failed, reason: " << status;
+            }
         }
```
Comment (satanson) [inline be/src/exprs/jit/jit_expr.cpp:127]: Have you override JITExpr::clone method?  it seems required.

both runtime filter and pushdown filter needs invoke this clone method

Comment (fzhedu) [reply be/src/exprs/jit/jit_expr.cpp:127]: yes in .h file
```
    Expr* clone(ObjectPool* pool) const override { return JITExpr::create(pool, _expr); }
```

File: be/src/exprs/jit/jit_engine.cpp [hunk 2]
```diff
@@ -181,14 +184,17 @@ void* JITEngine::compile_module(std::unique_ptr<llvm::Module> module, std::uniqu
...
-    _resource_ref_count_map[expr_name] = 1;
+    _resource_ref_count_map[expr_name] = 0;
     // Lookup the function in the JIT engine, this will trigger the compilation.
     return lookup_function(expr_name);
 }
 
 Status JITEngine::remove_module(const std::string& expr_name) {
     std::lock_guard<std::mutex> lock(_mutex);
-    if (--_resource_ref_count_map[expr_name] > 0) {
+    if (!_resource_ref_count_map.contains(expr_name)) {
```
Comment (silverbullet233) [inline be/src/exprs/jit/jit_engine.cpp:194]: add a DCHECK here?

**Review by satanson — APPROVED** [review]

File: be/src/exprs/jit/jit_engine.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exprs/jit/jit_engine.cpp:194]: ok, I will do that later.

**Review by silverbullet233 — APPROVED** [review]

**Pull request merged**